core/corehttp/gateway_handler.go

```diff
@@ -781,3 +808,33 @@ func preferred404Filename(acceptHeaders []string) (string, string, error) {
 
 	return "", "", fmt.Errorf("there is no 404 file for the requested content types")
 }
+
+// Attempt to fix redundant /ipfs/ namespace as long as resulting
+// 'intended' path is valid.  This is in case gremlins were tickled
+// wrong way and user ended up at /ipfs/ipfs/{cid} or /ipfs/ipns/{id}
+// like in bafybeien3m7mdn6imm425vc2s22erzyhbvk5n3ofzgikkhmdkh5cuqbpbq :^))
+func fixupSuperfluousNamespace(w http.ResponseWriter, urlPath string, urlQuery string) bool {
```

Comment:
So... I think we _also_ need to take the `prefix` into account. Users can (and sometimes do) mount the gateway under some other path.

Reply:
I don't believe the prefix thing adds any net value. Security/complexity price is too high imo.
See my proposal to remove support for X-Ipfs-Gateway-Prefix: https://github.com/ipfs/go-ipfs/issues/7702# – I forgot about it, if you give me :+1:  there I can remove it in a separate PR.

Reply:
I agree but can we have a bit of a deprecation period here?

1. Do our best to support it here.
2. Print big warnings for the next release cycle with a request to open an issue.
3. Remove it in the release cycle after that.

Reply:
Alternatively, _here_ we can just say "if prefix is non-empty, print an error and don't provide the nice redirect".

Reply:
Brave and other vendors don't rely on prefix, so I added a simple failsafe in https://github.com/ipfs/go-ipfs/pull/7930/commits/763a120ed28c6cd2e77077c6bb0281c583a6b580    (disables fixup when prefix is present). 

This is less code and will be easier to remove after prefix deprecation ends (we will track that in https://github.com/ipfs/go-ipfs/issues/7702#issuecomment-803175828).